"ActionInput not found for filename" exception crashing Bazel

### Description of the bug:


Very rarely, Bazel crashes with error messages like these:
```
14:39:02   211 FATAL: bazel crashed due to an internal error. Printing stack trace:
14:39:02   212 java.lang.RuntimeException: Unrecoverable error while evaluating node 'ActionLookupData0{actionLookupKey=ConfiguredTargetKey{label=//atlassian-embed-repo/atlassian-embed/src/main/kotlin/com/lucidchart/atlassian/model/lucid/documentembed:documentembed, config=BuildConfigurationKey[9e0af580f2bac6c86c12e854872b9f6763fe92a00d331709dbc4890bd7b17228]}, actionIndex=0}' (requested by nodes 'ActionLookupData1{actionLookupKey=ConfiguredTargetKey{label=//atlassian-embed-repo/atlassian-embed/src/main/kotlin/com/lucidchart/atlassian/model/lucid/documentembed:documentembed, config=BuildConfigurationKey[9e0af580f2bac6c86c12e854872b9f6763fe92a00d331709dbc4890bd7b17228]}, actionIndex=1}', 'ActionLookupData3{actionLookupKey=ConfiguredTargetKey{label=//atlassian-embed-repo/atlassian-embed/src/main/kotlin/com/lucidchart/atlassian/model/lucid/documentembed:documentembed, config=BuildConfigurationKey[9e0af580f2bac6c86c12e854872b9f6763fe92a00d331709dbc4890bd7b17228]}, actionIndex=3}')
14:39:02   213 	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:547)
14:39:02   214 	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:435)
14:39:02   215 	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.compute(Unknown Source)
14:39:02   216 	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.compute(Unknown Source)
14:39:02   217 	at java.base/java.util.concurrent.ForkJoinTask$InterruptibleTask.exec(Unknown Source)
14:39:02   218 	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
14:39:02   219 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
14:39:02   220 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
14:39:02   221 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
14:39:02   222 Caused by: java.lang.NullPointerException: ActionInput not found for filename bazel-out/k8-fastbuild/bin/atlassian-embed-repo/atlassian-embed/src/main/kotlin/com/lucidchart/atlassian/core/lucidproduct/lucidproduct.jar in CacheNotFoundException
14:39:02   223 	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:1008)
14:39:02   224 	at com.google.devtools.build.lib.remote.common.BulkTransferException.getLostInputs(BulkTransferException.java:92)
14:39:02   225 	at com.google.devtools.build.lib.remote.RemoteSpawnRunner.execLocallyAndUploadOrFail(RemoteSpawnRunner.java:591)
14:39:02   226 	at com.google.devtools.build.lib.remote.RemoteSpawnRunner.exec(RemoteSpawnRunner.java:346)
14:39:02   227 	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:161)
14:39:02   228 	at com.google.devtools.build.lib.dynamic.RemoteBranch.runRemotely(RemoteBranch.java:89)
14:39:02   229 	at com.google.devtools.build.lib.dynamic.RemoteBranch.callImpl(RemoteBranch.java:153)
14:39:02   230 	at com.google.devtools.build.lib.dynamic.Branch.call(Branch.java:188)
14:39:02   231 	at com.google.devtools.build.lib.dynamic.Branch.call(Branch.java:43)
14:39:02   232 	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:128)
14:39:02   233 	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:74)
14:39:02   234 	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:80)
14:39:02   235 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
14:39:02   236 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
14:39:02   237 	at java.base/java.lang.Thread.run(Unknown Source)
```

Here are a few more instances of the error occurring:
[error2.txt](https://github.com/user-attachments/files/22858146/error2.txt)
[error3.txt](https://github.com/user-attachments/files/22858145/error3.txt)
[error4.txt](https://github.com/user-attachments/files/22858147/error4.txt)

These errors are happening across different mnemonics, so we suspect it's a bug in Bazel or our remote execution software (Buildfarm), but we're not sure which.

We have dynamic execution enabled, in case that's relevant.

### Which category does this issue belong to?

Remote Execution

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.


Because the error occurs so rarely, it's very difficult to create a minimum reproducability case.

### Which operating system are you running Bazel on?


Ubuntu 24.04.3

### What is the output of `bazel info release`?


release 8.4.0

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.


_No response_

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?


```text

```

### If this is a regression, please try to identify the Bazel commit where the bug was introduced with bazelisk --bisect.


_No response_

### Have you found anything relevant by searching the web?

We've started a discussion in the Bazel Slack workspace:
https://bazelbuild.slack.com/archives/CD4MDG09Z/p1760125221906699

### Any other information, logs, or outputs that you want to share?

_No response_

--- a/src/main/java/com/google/devtools/build/lib/remote/RemoteSpawnRunner.java
+++ b/src/main/java/com/google/devtools/build/lib/remote/RemoteSpawnRunner.java
@@ -244,8 +244,7 @@ public class RemoteSpawnRunner implements SpawnRunner {
     } catch (CredentialHelperException e) {
       throw createExecExceptionForCredentialHelperException(e);
     } catch (IOException e) {
-      return execLocallyAndUploadOrFail(
-          action, spawn, context, uploadLocalResults, e, FailureReason.DOWNLOAD);
+      return execLocallyAndUploadOrFail(action, spawn, context, uploadLocalResults, e);
     }
 
     if (remoteOptions.remoteRequireCached) {
@@ -266,23 +265,16 @@ public class RemoteSpawnRunner implements SpawnRunner {
 
     AtomicBoolean useCachedResult = new AtomicBoolean(acceptCachedResult);
     AtomicBoolean forceUploadInput = new AtomicBoolean(false);
-    AtomicBoolean ioExceptionCausedByUpload = new AtomicBoolean(false);
     try {
       return retrier.execute(
           () -> {
-            ioExceptionCausedByUpload.set(false);
             // Upload the command and all the inputs into the remote cache.
             try (SilentCloseable c = prof.profile(UPLOAD_TIME, "upload missing inputs")) {
               Duration networkTimeStart = action.getNetworkTime().getDuration();
               Stopwatch uploadTime = Stopwatch.createStarted();
               // Upon retry, we force upload inputs
-              try {
-                remoteExecutionService.uploadInputsIfNotPresent(
-                    action, forceUploadInput.getAndSet(true));
-              } catch (IOException e) {
-                ioExceptionCausedByUpload.set(true);
-                throw e;
-              }
+              remoteExecutionService.uploadInputsIfNotPresent(
+                  action, forceUploadInput.getAndSet(true));
 
               // subtract network time consumed here to ensure wall clock during upload is not
               // double
@@ -352,13 +344,7 @@ public class RemoteSpawnRunner implements SpawnRunner {
     } catch (CredentialHelperException e) {
       throw createExecExceptionForCredentialHelperException(e);
     } catch (IOException e) {
-      return execLocallyAndUploadOrFail(
-          action,
-          spawn,
-          context,
-          uploadLocalResults,
-          e,
-          ioExceptionCausedByUpload.get() ? FailureReason.UPLOAD : FailureReason.DOWNLOAD);
+      return execLocallyAndUploadOrFail(action, spawn, context, uploadLocalResults, e);
     }
   }
 
@@ -580,8 +566,7 @@ public class RemoteSpawnRunner implements SpawnRunner {
       Spawn spawn,
       SpawnExecutionContext context,
       boolean uploadLocalResults,
-      IOException cause,
-      FailureReason reason)
+      IOException cause)
       throws ExecException, InterruptedException, IOException {
     // Regardless of cause, if we are interrupted, we should stop without displaying a user-visible
     // failure/stack trace.
@@ -589,9 +574,8 @@ public class RemoteSpawnRunner implements SpawnRunner {
       throw new InterruptedException();
     }
     // If the failure is caused by eviction of inputs to the current action that are only available
-    // remotely, try to regenerate the lost inputs. This doesn't make sense for outputs of the
-    // current action.
-    if (reason == FailureReason.UPLOAD && cause instanceof BulkTransferException e) {
+    // remotely, try to regenerate the lost inputs.
+    if (cause instanceof BulkTransferException e) {
       e.getLostArtifacts(context.getInputMetadataProvider()::getInput).throwIfNotEmpty();
     }
     if (remoteOptions.remoteLocalFallback && !RemoteRetrierUtils.causedByExecTimeout(cause)) {

--- a/src/main/java/com/google/devtools/build/lib/remote/common/BulkTransferException.java
+++ b/src/main/java/com/google/devtools/build/lib/remote/common/BulkTransferException.java
@@ -13,7 +13,6 @@
 // limitations under the License.
 package com.google.devtools.build.lib.remote.common;
 
-import static com.google.common.base.Preconditions.checkNotNull;
 import static com.google.common.collect.ImmutableList.toImmutableList;
 
 import com.google.common.base.Joiner;
@@ -93,10 +92,25 @@ public class BulkTransferException extends IOException {
       CacheNotFoundException e = (CacheNotFoundException) suppressed;
       var missingDigest = e.getMissingDigest();
       var execPath = e.getExecPath();
-      checkNotNull(execPath, "exec path not known for action input with digest %s", missingDigest);
+      if (execPath == null) {
+        // This can happen if the lost artifact is not an input of the action, but a special output
+        // such as stdout/stderr. This can't be solved by the rewinding that LostArtifacts would
+        // trigger, but is rather a failure of the current action execution.
+        if (e.getFilename() == null) {
+          throw new IllegalArgumentException(
+              "CacheNotFoundException that may represent a lost artifact should have been annotated"
+                  + " with a filename",
+              e);
+        }
+        return LostArtifacts.EMPTY;
+      }
       var actionInput = actionInputResolver.apply(execPath);
-      checkNotNull(
-          actionInput, "ActionInput not found for filename %s in CacheNotFoundException", execPath);
+      if (actionInput == null) {
+        // This can happen if the lost artifact is not an input of the action, but an output that
+        // e.g. failed to be retrieved from the remote cache after a cache hit. This also can't be
+        // solved by the rewinding that LostArtifacts would trigger.
+        return LostArtifacts.EMPTY;
+      }
       byDigestBuilder.put(DigestUtil.toString(missingDigest), actionInput);
     }
     var byDigest = byDigestBuilder.buildKeepingLast();

--- a/src/main/java/com/google/devtools/build/lib/remote/common/CacheNotFoundException.java
+++ b/src/main/java/com/google/devtools/build/lib/remote/common/CacheNotFoundException.java
@@ -62,6 +62,11 @@ public final class CacheNotFoundException extends IOException {
     return execPath;
   }
 
+  @Nullable
+  public String getFilename() {
+    return filename;
+  }
+
   @Override
   public String getMessage() {
     String message =

